src/braket/experimental/autoqasm/api.py

```diff
@@ -351,6 +351,20 @@ def _convert_subroutine(
                 _wrap_for_oqpy_subroutine(_clone_function(f), options)
             )
 
+            # todo: need to validate in other codepaths too?
+            # Validate and declare used qubits
+            quantum_indices = {
+                i
+                # todo: is there a better way to access argument types?
```

Comment:
`aq.gate` and `aq.gate_calibration` use `aq_program.GateArgs()` to separate qubit and non-qubit argument. It only separates qubit and float types there. I think subroutines consider more than these two types. One idea is we could generalize `GateArgs` to include more types beyond just qubit and float.

Reply:
Yes, I see. That sounds like a good path to me

Reply:
One difference is that `GateArgs` allows arbitrary ordering of the qubit and angle args in Python, and then separates them out as required for OQ3 gate definitions. For subroutines we want the order in Python and the order in OQ3 to match exactly (right? can anyone think of a counterexample to this?).

Reply:
Yes, though here we do want to filter out only the qubit args.